lib/spack/spack/solver/concretize.lp

```diff
@@ -19,30 +19,30 @@
-:- attr("node_flag_compiler_default", PackageNode), not attr("node", PackageNode).
-:- attr("node_flag", PackageNode, _, _), not attr("node", PackageNode).
-:- attr("node_flag_source", PackageNode, _, _), not attr("node", PackageNode).
-:- attr("no_flags", PackageNode, _), not attr("node", PackageNode).
-:- attr("external_spec_selected", PackageNode, _), not attr("node", PackageNode).
-:- attr("depends_on", ParentNode, _, _), not attr("node", ParentNode).
-:- attr("depends_on", _, ChildNode, _), not attr("node", ChildNode).
-:- attr("node_flag_source", ParentNode, _, _), not attr("node", ParentNode).
-:- attr("node_flag_source", _, _, ChildNode), not attr("node", ChildNode).
-
-:- attr("virtual_node", VirtualNode), not provider(_, VirtualNode).
-:- provider(_, VirtualNode), not attr("virtual_node", VirtualNode).
-:- provider(PackageNode, _), not attr("node", PackageNode).
-
-:- attr("root", node(ID, PackageNode)), ID > min_dupe_id.
+error(100, "e5") :- attr("root", PackageNode), not attr("node", PackageNode).
+error(100, "e6") :- attr("version", PackageNode), not attr("node", PackageNode).
+error(100, "e7") :- attr("node_version_satisfies", PackageNode), not attr("node", PackageNode).
+error(100, "e7") :- attr("hash", PackageNode, _), not attr("node", PackageNode).
+error(100, "e8") :- attr("node_platform", PackageNode, _), not attr("node", PackageNode).
+error(100, "e9") :- attr("node_os", PackageNode, _), not attr("node", PackageNode).
+error(100, "e10") :- attr("node_target", PackageNode, _), not attr("node", PackageNode).
+error(100, "e11") :- attr("node_compiler_version", PackageNode, _, _), not attr("node", PackageNode).
+error(100, "e12") :- attr("variant_value", PackageNode, _, _), not attr("node", PackageNode).
+error(100, "e13") :- attr("node_flag_compiler_default", PackageNode), not attr("node", PackageNode).
+error(100, "e14") :- attr("node_flag", PackageNode, _, _), not attr("node", PackageNode).
+error(100, "e15") :- attr("node_flag_source", PackageNode, _, _), not attr("node", PackageNode).
+error(100, "e16") :- attr("no_flags", PackageNode, _), not attr("node", PackageNode).
+error(100, "e17") :- attr("external_spec_selected", PackageNode, _), not attr("node", PackageNode).
+error(100, "e18") :- attr("depends_on", ParentNode, _, _), not attr("node", ParentNode).
+error(100, "e19") :- attr("depends_on", _, ChildNode, _), not attr("node", ChildNode).
+error(100, "e20") :- attr("node_flag_source", ParentNode, _, _), not attr("node", ParentNode).
+error(100, "e21") :- attr("node_flag_source", _, _, ChildNode), not attr("node", ChildNode).
+
+error(100, "e1") :- attr("virtual_node", VirtualNode), not provider(_, VirtualNode).
+error(100, "e2") :- provider(_, VirtualNode), not attr("virtual_node", VirtualNode).
+error(100, "e3") :- provider(PackageNode, _), not attr("node", PackageNode).
+
+error(100, "e4") :- attr("root", node(ID, PackageNode)), ID > min_dupe_id.
```

Comment:
Introducing these error facts helped me figure out why 

```
packages:
  virtual-w:
    require: cflags="-O1"
```

(i.e. requirements from the virtual that do not explicitly mention an implementation): virtuals can't propagate most constraints to nodes.